fabric-common/lib/Utils.js

```diff
@@ -447,26 +447,31 @@ module.exports.getDefaultKeyStorePath = () => {
 	return path.join(os.homedir(), '.hfc-key-store');
 };
 
-const CryptoKeyStore = function (KVSImplClass, opts) {
-	let store;
-
-	this._getKeyStore = async function () {
-		if (!store) {
-			const newInstance = require('./impl/CryptoKeyStore');
-			store = await newInstance(KVSImplClass, opts);
-			await store.initialize();
-		}
-
-		return store;
-	};
-};
-
+/**
+ *
+ * @param {function} [KVSImplClass] Optional. The built-in key store saves private keys.
+ *    The key store may be backed by different {@link KeyValueStore} implementations.
+ *    If specified, the value of the argument must point to a module implementing the
+ *    KeyValueStore interface.
+ * @param {Object} [opts] Implementation-specific option object used in the constructor
+ * @return {CryptoKeyStore}
+ */
 module.exports.newCryptoKeyStore = (KVSImplClass, opts) => {
-	// this function supports skipping any of the arguments such that it can be called in any of the following fashions:
 	// - newCryptoKeyStore(CouchDBKeyValueStore, {name: 'member_db', url: 'http://localhost:5984'})
-	// - newCryptoKeyStore({path: '/tmp/app-state-store'})
 	// - newCryptoKeyStore()
-	return new CryptoKeyStore(KVSImplClass, opts);
+
+	const kvs = typeof KVSImplClass === 'function' ? new KVSImplClass(opts) : new InMemoryKeyValueStore();
+	const CryptoKeyStore = require('./impl/CryptoKeyStore');
+	const cks = new CryptoKeyStore(kvs);
+	cks._getKeyStore = async () => {
+		if (!cks.isReady) {
```

Comment:
Does this mean that `cks.initialize()` might be called more than once if another call of `cks._getKeyStore()` passes this line before a previous call completes the await on `cks.initialize()` on the following line and sets `cks.isReady` to `true`? I wonder if something like this might be more reliable:
```
const cks = new CryptoKeyStore(kvs);
const isReady = cks.initialize();
cks._getKeyStore = async () => {
    await isReady;
    return cks;
};
```
It does mean `cks.initialize()` gets kicked off before anyone calls `cks._getKeyStore()` though. I don't know if that is a problem.

Reply:
Now I think about it, I'm not really sure what the point is of having the `cks._getKeyStore()` at all, since it just calls initialize() and returns `cks`. It seems to me that it might be simpler to have `newCryptoKeyStore` as an async factory function something like:
```
module.exports.newCryptoKeyStore = async (KVSImplClass, opts) => {
    const kvs = typeof KVSImplClass === 'function' ? new KVSImplClass(opts) : new InMemoryKeyValueStore();
    const CryptoKeyStore = require('./impl/CryptoKeyStore');
    const cks = new CryptoKeyStore(kvs);
    await cks.initialize();
    return cks;
};
```
Maybe it's just to avoid changing `newCryptoKeyStore` to an async function and requiring calling code changes?

Even better might be to push this async factory function into the CryptoKeyStore module and export that instead of the CryptoKeyStore class itself so calling code can't get it wrong and fail to initialize a new CryptoKeyStore.

Reply:
Reply to “Maybe it's just to avoid changing newCryptoKeyStore to an async function and requiring calling code changes?”
Yes.

But for the last section: 
"Even better might be to push this async factory function into the CryptoKeyStore module and export that instead of the CryptoKeyStore class itself so calling code can't get it wrong and fail to initialize a new CryptoKeyStore."

I am not sure I get the idea. Do you mean we adding a new static method inside `CryptoKeyStore` such as 
```
async static newCryptoKeyStore (kvs) {
  
    const cks = new CryptoKeyStore(kvs);
    await cks.initialize();
    return cks;
}
```

Reply:
I was thinking it could be a module-scope function. Then only that function could be exported and the actual CryptoKeyStore class could be entirely private within the module (not exported). So the only way to access the CryptoKeyStore would be by calling `newCryptoKeyStore()` and there would be no possible way (without subverting the Node module system) for any calling code to fail to initialize it correctly.

That does assume that no other code needs to directly access the CryptoKeyStore class though, such as to extend it. I am not sure whether or not that is the case.

Reply:
> I was thinking it could be a module-scope function. Then only that function could be exported and the actual CryptoKeyStore class could be entirely private within the module (not exported). So the only way to access the CryptoKeyStore would be by calling `newCryptoKeyStore()` and there would be no possible way (without subverting the Node module system) for any calling code to fail to initialize it correctly.
> 
> That does assume that no other code needs to directly access the CryptoKeyStore class though, such as to extend it. I am not sure whether or not that is the case.

Gotcha, but I prefer to keeping the class exported, based on idea that it is developer's responsibility to handle `initialize` action (when and how), while we also provide a method to get state-ready instance via utils.js.